src/group_theory/free_product.lean

```diff
@@ -676,4 +677,146 @@ instance {ι : Type*} (G : ι → Type*) [∀ i, group (G i)] [hG : ∀ i, is_fr
     { intros g hfg, ext i x, simpa using hfg ⟨i, x⟩, }
   end, }
 
+/-- A free group is a free product of copies of the free_group over one generator. -/
+
+-- NB: One might expect this theorem to be phrased with ℤ, but ℤ is an additive group,
+-- and using `multiplicative ℤ` runs into diamond issues.
+@[simps]
+def _root_.free_group_equiv_free_product {ι : Type u_1} :
+  free_group ι ≃* free_product (λ (_ : ι), free_group unit) :=
+begin
+  refine monoid_hom.to_mul_equiv _ _ _ _,
+  exact free_group.lift (λ i, @free_product.of ι _ _ i (free_group.of unit.star)),
+  exact free_product.lift (λ i, free_group.lift (λ pstar, free_group.of i)),
+  { ext i, refl, },
+  { ext i a, cases a, refl, },
+end
+
+section ping_pong_lemma
+
+open_locale pointwise
+open_locale cardinal
+
+variables [nonempty ι]
+variables [nontrivial ι]
+variables {G : Type u_1} [group G]
+variables (a : ι → G)
+
+-- A group action on α, and the ping-pong sets
+variables {α : Type*} [mul_action G α]
+variables (X : ι → set α)
+variables (Y : ι → set α)
+variables (hXnonempty : ∀ i, (X i).nonempty)
```

Comment:
I think I would squeeze a bit more of these on 1 line. But it's not very important.

Reply:
Actually, looking at it it seems silly to have both `[nontrivial]` and `[nonempty]`. I’ll fix that.